packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -528,12 +499,108 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+            _closeShortWithAmountsFromLp(
+                data.vaultId,
+                wPowerPerpAmountInLp,
+                data.wPowerPerpAmountToBurn,
+                data.collateralToWithdraw.add(data.collateralToFlashloan),
+                data.limitPriceEthPerPowerPerp
+            );
+            ControllerHelperUtil.checkPartialLpClose(nonfungiblePositionManager, controller, data.vaultId, data.tokenId, data.liquidityPercentage);
+        } 
+        else if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_REBALANCE_VAULT_NFT) {
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            RebalanceVaultNftParams[] memory data = abi.decode(_calldata, (RebalanceVaultNftParams[]));
+            console.log(data.length);
+
+            for(uint256 i; i < data.length; i++) {
+                if (data[i].rebalanceVaultNftType == RebalanceVaultNftType.IncreaseLiquidity) {
+                    IController(controller).deposit{value: _amount}(data[i].vaultId);
+                    IController(controller).withdrawUniPositionToken(data[i].vaultId);
+
+                    IncreaseLiquidityParam memory increaseLiquidityParam = abi.decode(data[i].data, (IncreaseLiquidityParam));
+
+                    if (increaseLiquidityParam.wPowerPerpAmountToMint > 0) {
+                        IController(controller).mintWPowerPerpAmount{value: increaseLiquidityParam.collateralToDeposit}(
+                            data[i].vaultId,
+                            increaseLiquidityParam.wPowerPerpAmountToMint,
+                            0
+                        );
+                    }
+
+                    INonfungiblePositionManager.IncreaseLiquidityParams memory _params = INonfungiblePositionManager.IncreaseLiquidityParams({
+                        tokenId: increaseLiquidityParam.tokenId,
+                        amount0Desired: (isWethToken0) ? increaseLiquidityParam.wethAmountToLp : increaseLiquidityParam.wPowerPerpAmountToMint,
+                        amount1Desired: (isWethToken0) ? increaseLiquidityParam.wPowerPerpAmountToMint : increaseLiquidityParam.wethAmountToLp,
+                        amount0Min: increaseLiquidityParam.amount0Min,
+                        amount1Min: increaseLiquidityParam.amount1Min,
+                        deadline: block.timestamp
+                    });
+                } else if (data[i].rebalanceVaultNftType == RebalanceVaultNftType.RepayFlashloan) {
+                    // deposit Uni NFT token in vault
+                    INonfungiblePositionManager(nonfungiblePositionManager).approve(controller, data[i].tokenId);
+                    IController(controller).depositUniPositionToken(data[i].vaultId, data[i].tokenId);
+
+                    // remove flashloan amount in ETH from vault + any amount of collateral user want to withdraw (sum <= vault.collateralAmount)
+                    IController(controller).withdraw(data[i].vaultId, _amount);
```

Comment:
think we need to call checkLpMintExcess() here as there might be excess wPowerPerp that is not used in the LP